kernel/inductive.ml

```diff
@@ -1111,6 +1113,88 @@ let filter_stack_domain env nr p stack =
   in
   filter_stack env ar stack
 
+let merge_filter l l' =
+  let n = List.length l in
+  let n' = List.length l' in
+  if n < n' then List.map2 (&&) l (List.firstn n l')
+  else List.map2 (&&) (List.firstn n' l) l'
+
+let find_uniform_parameters recindx nargs bodies =
+  let nbodies = Array.length bodies in
+  let min_indx = Array.fold_left min nargs recindx in
+  let argsign = List.make min_indx true in
+  (* We work only on the i-th body but are in the context of n bodies *)
+  let rec aux i k argsign c =
+    let f, l = decompose_app_list c in
+    match kind f with
+    | Rel n ->
+      (* A recursive reference to the i-th body *)
+      if Int.equal n (nbodies + k - i) then
+        merge_filter (List.map_i (fun j a ->
+            match kind a with
+            | Rel m when Int.equal m (k - j) ->
+              (* a reference to the j-th parameter *)
+              true
+            | _ ->
+              false) 0 l) argsign
+      else
+        argsign
+    | _ -> fold_constr_with_binders succ (aux i) k argsign c
+  in
+  Array.fold_left_i (fun i -> aux i 0) argsign bodies
+
+let drop_uniform_parameters argsign bodies =
+  let nbodies = Array.length bodies in
+  let nargs = List.length argsign in
+  let rec aux i k c =
+    let f, l = decompose_app_list c in
+    match kind f with
+    | Rel n ->
+      (* A recursive reference to the i-th body *)
+      if Int.equal n (nbodies + k - i) then
+        let args, extra = List.chop nargs l in
+        let new_args = snd (List.filter2 (fun uniform _ -> not uniform) argsign args) in
+        Term.applist (f, new_args @ extra)
```

Comment:
This term could be ill-typed, couldn't it? How do we know that uniform parameters cannot depend on a non-uniform one? With SProp we lose injectivity of dependence on variables.

Reply:
You're right, as in:
```
Inductive sTrue : SProp := sI.
Parameter P : sTrue -> Prop.
Check fix f a (e : P a) n :=
  match n with
  | 0 => 0
  | S n => f sI e n
  end.
```
And the same would typically also happen if we had η on `unit`.

Then we could either expect the uniform arguments to be a continuous prefix of the arguments, as that would cover the interesting cases, or to consider that definitionally-inhabited subsingleton arguments can always be extruded?

Characterizing the latter does not look so easy, so, by simplicity, I would go towards the first solution. Ok for you? (That means reducing the data `argsign` to just an `int`.)

Reply:
Actually, I don't see anything wrong with extruding definitionally-inhabited subsingleton arguments, or more generally with uniform arguments that do not constitute a prefix of the arguments. Assuming an interleaving of uniform `γ` and `γ'` and non-uniform `δ` as in:
```coq
fix f γ δ γ' n := φ(f γ δ' γ', ..., f γ δ'' γ')
```
we can translate into:
```coq
fun γ δ γ' => (fix f δ n := φ(f δ', ..., f δ'')) δ
```
The code of the PR does not explicitly do neither a `fun γ γ' => ...` extrusion (which might be ill-typed in general) nor a `fun γ δ γ' => ... δ` mixed extrusion/expansion, but what `check_nested_fix_body` is doing is compatible with both: it checks the guard of `φ(f δ', ..., f δ'')` in the context `γ δ γ'` where `γ` and `γ'` inherit the size of their argument in the stack (if any).

Reply:
I still fail to see why I (x, y, z) is a well-formed type in your comment. The z variable has not the right type as it has type C{x, y₀} rather than C{x, y} where y₀ is the outermost y variable.

Reply:
Regarding typing, I'm finding a small weakness in the algorithm: the arguments could be normalized before checking their uniformity: otherwise we may find `u` different of `y` but convertible to `y`, and fail to extrude it in a way which respects typing.

Assuming the normalization done, we know `z:C{x, y}` and `z:C{x, u}` from the original term and thus `C{x, y}` and `C{x, u}` convertible. Since `u` is not convertible with `y`, the convertibility proof of `C{x, y} ≣ C{x, u}` either eventually discard `y` and `u` or make them convertible from some another reason (e.g. being in an SProp). In any cases, the idea is that the convertibility proof of `C{x, y} ≣ C{x, u}` can be modified into a proof of convertibility of `C{x, y} ≣ C{x, y₀}`.

I don't think that missing an extrusion because of some `u` not detected as being convertible to the `y` is bad as we would just be more restrictive in the guard compared what would give the optimal extrusion. However, do you think e.g. it would be better to check the convertibility of `u` with `y` (thus taking into account strict propness), rather than just reasoning on the form?

Reply:
Would it be wrong to do a proof by induction on the structure and the reducts of `C{x, _}`? Then, am I missing other base cases than when `C{_,_}` does not depend on its second argument and when both `C{x, y}` and `C{x, u}` are detected proof irrelevant? (I'm asking because I'm not so familiar with the rules for proof irrelevance, so I may be missing some rules.)

More generally, is your intuition telling you that the result is wrong or is it that you'd like, as a rule, that statements about Coq TT are always formalized? In the latter case, I would need help because I don't have an infrastructure such as MetaCoq at hand to do a computer-checked proof.

Reply:
Alternatively, we can also renounce to non-prefix extrusion. This would still certainly cover most of the interesting cases.

Reply:
I'd prefer not to implement non-prefix extrusion so as to err on the safe side.

Reply:
Ok, done.

I also added more tests as well as found and fixed a residual bug (wrongly dropping the "NeedReduce" status of the uniform arguments of an inner fix by testing the need for reduction of only `extra_stack` in the previous version of the PR).